src/wpublish/QueriesAndChunks.cc

```diff
@@ -489,105 +499,168 @@ QueriesAndChunks::ScanTableSumsMap QueriesAndChunks::_calcScanTableSums() {
+    // Sum of Tasks in
+    int uqBootedTasksSum = 0;
+    int maxDarkQueries = 5 * _maxTasksBooted;
+    if (bootedTaskCount > maxDarkQueries) {
+        for (auto const& countQId : countQIdVect) {
+            // If enough UserQueries have been booted to get under the Task limit,
+            // stop booting UserQueries.
+            bool enoughTasksBooted = (bootedTaskCount - uqBootedTasksSum) <= maxDarkQueries;
+            LOGS(_log, LOG_LVL_DEBUG,
+                 __func__ << " a uq=" << countQId.qId << " enough=" << enoughTasksBooted
+                          << " bootedTaskCount=" << bootedTaskCount
+                          << " uqBootedTasksSum=" << uqBootedTasksSum
+                          << " sub=" << (bootedTaskCount - uqBootedTasksSum) << " max=" << maxDarkQueries);
+            if (enoughTasksBooted) {
+                break;
+            }
 
-/// Add a Task to the user query statistics.
-void QueryStatistics::addTask(wbase::Task::Ptr const& task) {
-    lock_guard<mutex> guard(_qStatsMtx);
-    if (queryId != task->getQueryId()) {
-        string const msg = "QueryStatistics::" + string(__func__) +
-                           ": the task has queryId=" + to_string(task->getQueryId()) +
-                           " where queryId=" + to_string(queryId) + " was expected.";
-        throw util::Bug(ERR_LOC, msg);
+            // Get query info
+            QueryId qIdToBoot = countQId.qId;
+            QueryStatistics::Ptr queryToBoot = getStats(qIdToBoot);
+            if (queryToBoot != nullptr) {
+                if (_bootUserQuery(queryToBoot, bSched)) {
+                    uqBootedTasksSum += countQId.count;
+                    LOGS(_log, LOG_LVL_DEBUG,
+                         __func__ << " a booted uq=" << countQId.qId << ":count=" << countQId.count
+                                  << " uqBootedTasksSum=" << uqBootedTasksSum);
+                }
+            } else {
+                // This really should never happen.
+                LOGS(_log, LOG_LVL_ERROR, __func__ << " queryToBoot is nullptr qIdToBoot=" << countQId.qId);
+            }
+        }
     }
-    _tasks.push_back(task);
-}
 
-/// @return the number of Tasks that have been booted for this user query.
-int QueryStatistics::getTasksBooted() {
-    lock_guard<mutex> guard(_qStatsMtx);
-    return _tasksBooted;
-}
+    // Last check, if any of the queries have several booted `Task`s but have not been booted,
+    // they should be booted now.
+    for (auto const& countQId : countQIdVect) {
+        QueryId qIdToCheck = countQId.qId;
+        // Get query info
+        QueryStatistics::Ptr queryToCheck = getStats(qIdToCheck);
+        if (queryToCheck == nullptr) {
+            LOGS(_log, LOG_LVL_ERROR, __func__ << " Couldn't locate qIdToCheck=" << qIdToCheck);
+            // This really should never happen, but try to boot the next query in the list.
+            continue;
```

Comment:
Then the code should throw an exception to indicate a bug:
```
util::Bug(ERR_LOC, ...);
```

Reply:
I can only see this possibly happening if something is cancelled, and in that case, it shouldn't crash the worker.